test(cypress): add Cypress tests for workflow pages

lorenzo-cavazzi opened the pull request:
Add Cypress tests to cover workflows.

fix #2852
/deploy #persist #notest

ciyer commented on cypress-tests/cypress/e2e/useSession.cy.ts:
```diff
@@ -109,23 +115,54 @@ describe("Basic public project functionality", () => {
+      cy.get(".xterm-helper-textarea")
+        .click()
+        .type(`renku run --name ${workflow.name} echo "123" > ${workflow.output}{enter}`);
+      cy.get("#filebrowser").should("be.visible").contains(workflow.output).should("be.visible");
+      cy.get("#jp-git-sessions")
+        .get(`button[title="Push committed changes (ahead by 1 commits)"]`)
+        .should("not.exist");
+
+      // push the changes
+      cy.get(`[data-id="jp-git-sessions"]`).should("be.visible").click();
+      cy.get("#jp-git-sessions").contains(projectTestConfig.projectName).should("be.visible");
+      cy.get("#jp-git-sessions")
+        .get(`button[title="Push committed changes (ahead by 1 commits)"]`)
```
Wouldn't it be better to use the save feature that we add to sessions?

<img width="1072" alt="image" src="https://github.com/SwissDataScienceCenter/renku/assets/1196411/4596e0e3-797f-4462-b314-0aa7f041658c">

lorenzo-cavazzi replied:
Yes, that would be better. Actually, I remember now why I didn't go that way; there is bug currently causing an error when trying to use `renku save` from the sidecar API
https://github.com/SwissDataScienceCenter/renku-notebooks/issues/1575

![image](https://github.com/SwissDataScienceCenter/renku/assets/43481553/dbbdba27-0aa1-46e9-a8bb-8c693869334c)

lorenzo-cavazzi replied:
I added a stub of the code inline and a comment to update the logic there once the renku-notebooks issue is fixed

ciyer commented on cypress-tests/cypress/e2e/checkWorkflows.cy.ts:
```diff
@@ -0,0 +1,124 @@
+    cy.get("[data-cy=workflows-page]", { timeout: TIMEOUTS.long }).should(
+      "be.visible"
+    );
+    cy.dataCy("workflows-browser")
+      .should("be.visible")
+      .children()
+      .should("have.length", 9);
+    cy.dataCy("workflows-browser")
+      .children()
+      .contains("useless-workflow-with-kw")
+      .should("be.visible")
+      .click();
```
I think this is a bit easier to follow without reselecting the `workflows-browser`

```suggestion
    cy.dataCy("workflows-browser")
      .should("be.visible")
      .children()
      .should("have.length", 9)
      .contains("useless-workflow-with-kw")
      .should("be.visible")
      .click();
```

ciyer commented on cypress-tests/cypress/e2e/checkWorkflows.cy.ts:
```diff
@@ -0,0 +1,124 @@
+      .should("be.visible")
+      .click();
+
+    cy.dataCy("workflow-details").should("be.visible");
+    cy.dataCy("workflow-details")
+      .contains("This is defintely a useless workflow.")
+      .should("be.visible");
```
See above
```suggestion
    cy.dataCy("workflow-details").should("be.visible")
      .contains("This is defintely a useless workflow.")
      .should("be.visible");
```

ciyer commented on cypress-tests/cypress/e2e/checkWorkflows.cy.ts:
```diff
@@ -0,0 +1,124 @@
+      .should("be.visible");
+
+    // Play with the workflow browser and check composite workflows link other workflows
+    cy.dataCy("workflows-browser").should("be.visible");
+    cy.dataCy("workflows-browser").children().should("have.length", 9);
```
See above
```suggestion
    cy.dataCy("workflows-browser").should("be.visible").children().should("have.length", 9);
```

ciyer commented on cypress-tests/cypress/e2e/checkWorkflows.cy.ts:
```diff
@@ -0,0 +1,124 @@
+      .get(".rk-tree-item--children")
+      .should("be.visible");
+
+    cy.dataCy("workflow-details")
+      .contains("Workflow (Composite)")
+      .should("be.visible");
+    cy.dataCy("workflow-details")
+      .contains("div h5", "composite1")
+      .should("be.visible")
+      .click();
+
+    cy.dataCy("workflow-details")
+      .contains("renku workflow execute composite1")
+      .should("be.visible");
+    cy.dataCy("workflow-details")
+      .contains("Workflow (Composite)")
+      .should("be.visible");
+    cy.dataCy("workflow-details")
+      .contains("step1_input-1")
+      .should("be.visible");
+    cy.dataCy("workflow-details")
+      .contains("div h5", "so-many-steps")
+      .should("be.visible")
+      .click();
```
```suggestion
    cy.dataCy("workflow-details").within(() => {
      cy.root().contains("Workflow (Composite)").should("be.visible");
      cy.root().contains("div h5", "composite1").should("be.visible").click();
    });

    cy.dataCy("workflow-details").within(() => {
      cy.root()
        .contains("renku workflow execute composite1")
        .should("be.visible");
      cy.root().contains("Workflow (Composite)").should("be.visible");
      cy.root().contains("step1_input-1").should("be.visible");
      cy.root()
        .contains("div h5", "so-many-steps")
        .should("be.visible")
        .click();
    });
```

ciyer reviewed (changes requested):
I'm looking forward to having a test for workflows -- this will help us develop the UI more effectively.

I have some suggestions for changes to the tests that improve readability, but you can ignore them if you disagree.

lorenzo-cavazzi commented:
@ciyer thank you for the review. I incorporated the suggestions for simplifying the logic. I also added a few lines at the end to check from the Workflows tab the workflow ran properly.
I couldn't switch to using the Save session button cause it's currently broken (see inline) but we can switch to that later

ciyer reviewed (approved):
🎉
